lib/AST/ASTStructuralEquivalence.cpp

```diff
@@ -102,11 +102,103 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
 static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
                                      const TemplateArgument &Arg1,
                                      const TemplateArgument &Arg2);
+static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
+                                     const TemplateArgumentList &ArgL1,
+                                     const TemplateArgumentList &ArgL2);
 static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
                                      NestedNameSpecifier *NNS1,
                                      NestedNameSpecifier *NNS2);
 static bool IsStructurallyEquivalent(const IdentifierInfo *Name1,
                                      const IdentifierInfo *Name2);
+static bool IsStructurallyEquivalent(const DeclarationName Name1,
+                                     const DeclarationName Name2);
+
+using ContextVector = llvm::SmallVector<const DeclContext *, 4>;
+
+static void GetContexts(ContextVector &Contexts, const NamedDecl *ND) {
+  const DeclContext *Ctx = ND->getDeclContext();
+
+  // For ObjC methods, look through categories and use the interface as context.
+  if (auto *MD = dyn_cast<ObjCMethodDecl>(ND))
+    if (auto *ID = MD->getClassInterface())
+      Ctx = ID;
+
+  while (Ctx && Ctx->isFunctionOrMethod())
+    Ctx = Ctx->getParent();
+
+  // Collect named contexts.
+  while (Ctx) {
+    if (isa<NamedDecl>(Ctx))
+      Contexts.push_back(Ctx);
+    Ctx = Ctx->getParent();
+  }
+}
+
+static bool IsEquivalentContext(StructuralEquivalenceContext &Context, const ContextVector &Contexts1, const ContextVector &Contexts2) {
```

Comment:
I am afraid that we are going to miss too much cases here by implementing this logic all from scratch. A similar logic is already implemented in either NamedDecl::getQualifiedNameAsString or by index::generateUSRForDec. I think we should reuse the USR version ... see my other comment too

Reply:
The USR seems to be usable but again a string-like (slow?) check.

Reply:
I am not sure, I think we should maybe try both solutions. 

I like the context vector based solution you have and maybe that is the winner because seems much faster. I am just afraid that we are going to miss something or make errors when implementing that. We can't make big mistakes however when comparing strings.